vmm/src/api/http/http_endpoint.rs

```diff
@@ -50,6 +52,162 @@ use crate::config::RestoreConfig;
+    use std::fs::File;
+    use std::os::fd::IntoRawFd;
+
+    use crate::api::http::HttpError;
+
+    /// Abstraction over configuration types received via the HTTP API that
+    /// logically hold FDs. Helper for [`apply_new_fds_to_cfg`].
+    ///
+    /// Example: virtio-net device with externally opened FDs for the Tap device
+    pub trait ConfigWithFDs {
+        /// Returns the ID of the device.
+        ///
+        /// Used for logging.
+        fn id(&self) -> Option<&str>;
+
+        /// Returns how many FDs this type wants to own.
+        ///
+        /// `None` instead `Some(0)` indicates that this type wants to take
+        /// ownership of all available FDs. This is the case for config options
+        /// that don't know (yet) how many FDs belong to them, such as on hot
+        /// device attach. In these cases, only one single config object is
+        /// handled by the API call.
+        fn expected_num_fds(&self) -> Option<usize>;
+
+        /// Returns any FDs provided in the HTTP body.
+        ///
+        /// They will always be invalid. Used for logging.
+        fn fds_from_http_body(&self) -> Option<&[i32]>;
+
+        /// Updates the internal fds field to a new value.
+        fn set_fds(&mut self, fds: Option<Vec<i32>>);
+    }
+
+    mod config_with_fds_impls {
+        use super::ConfigWithFDs;
+        use crate::config::RestoredNetConfig;
+        use crate::vm_config::NetConfig;
+
+        impl ConfigWithFDs for NetConfig {
+            fn id(&self) -> Option<&str> {
+                self.id.as_deref()
+            }
+
+            fn expected_num_fds(&self) -> Option<usize> {
+                None
+            }
+
+            fn fds_from_http_body(&self) -> Option<&[i32]> {
+                self.fds.as_deref()
+            }
+
+            fn set_fds(&mut self, fds: Option<Vec<i32>>) {
+                self.fds = fds;
+            }
+        }
+
+        impl ConfigWithFDs for RestoredNetConfig {
+            fn id(&self) -> Option<&str> {
+                Some(self.id.as_str())
+            }
+
+            fn expected_num_fds(&self) -> Option<usize> {
+                Some(self.num_fds)
+            }
+
+            fn fds_from_http_body(&self) -> Option<&[i32]> {
+                self.fds.as_deref()
+            }
+
+            fn set_fds(&mut self, fds: Option<Vec<i32>>) {
+                self.fds = fds;
+            }
+        }
+    }
+
+    /// Applies FDs to configs for their corresponding devices, as part of the special
+    /// handling for devices backed by externally provided FDs.
+    ///
+    /// The FDs (via `files`) **must** be provided in the exact order matching the
+    /// config struct they belong to.
+    ///
+    /// See [module description] for more info.
+    ///
+    /// [module description]: self
+    pub fn apply_new_fds_to_cfg<T: ConfigWithFDs>(
+        // List of new files (well, actually FDs) that back up the device(s) of the given config(s).
+        files: Vec<File>,
```

Comment:
I think we could name this better (`device_fds`), and then avoid the need for an explanatory comment.

Reply:
fixed